[AIX] PGO codegen changes for function-sections.

mandlebug opened the pull request:
The existing PGO implementation creates a dependence from the profiling data symbol to the function descriptors for every function defined within the compilation unit. This forces the binder to include either all or no functions from an object file when linking a profile-generate enabled object file, which can break links that require the aggressive "include only whats used" link semantics that XCOFF provides. To remedy this we break up the profiling data symbol into per-function CSECTs by using the rename directive to emit the profd symbol for a single function to a unique CSECT, then rename it to the common name afterwards. We also have to split the counters into per-function CSECTs as they need to have a .ref relocation from the  counter section which is referenced by the functions, to the profiling-data symbol which is otherwise unreferenced.

mandlebug committed:
[AIX] PGO codegen changes for function-sections.

The existing PGO implementation creates a dependence from the profiling
data symbol to the function descriptors for every function defined
within the compilation unit. This forces the binder to include either all
or no functions from an object file when linking a profile-generate
enabled object file, which can break links that require the aggressive
"include only whats used" link semantics that XCOFF provides. To remedy
this we break up the profiling data symbol into per-function CSECTs by
using the rename directive to emit the profd symbol for a single function
to a unique CSECT, then rename it to the common name afterwards. We also
have to split the counters into per-function CSECTs as they need to have a
.ref relocation from the  counter section which is referenced by the
functions, to the profiling-data symbol which is otherwise unreferenced.

mandlebug committed:
Address first round of review comments.

- Fix spelling mistake in comment.
- Use llvm functions to get the string prefixes for pgo data and
  counter objects instead of hardcoding them.
- Remove some unused declarations from one of the test files.

amy-kwan reviewed (commented):
Just some minor nit comments.

mandlebug commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -2810,6 +2826,57 @@ void PPCAIXAsmPrinter::emitGlobalVariableHelper(const GlobalVariable *GV) {
+    // between the counters label and the label for the start of this structure.
+    // Use the subtract expression to get the GlobalValue for the counters
+    // global.
+    assert(GV->hasInitializer() &&
```
I really dislike this part because we are so tightly coupled to the implementation that if anyone changed the initializer format they would have to simultaneously modify the PPC target which is a problem. For ELF the section keys on the Comdats carry the relationship info in the IR implicitly.  

What do you think of an alternative approach where we could use a metadata node to explicitly attach to dependancy info to the data symbol? https://llvm.org/docs/LangRef.html#associated-metadata is a similar idea for ELF.

petrhosek replied:
I agree, this is really brittle and may force developers changing `InstrProfiling` to also change the `PowerPC` backend and they may not have the necessary expertise or a way to test their changes.

petrhosek commented on llvm/lib/Target/PowerPC/PPCAsmPrinter.cpp:
```diff
@@ -2960,6 +3027,54 @@ void PPCAIXAsmPrinter::emitPGORefs(Module &M) {
+
+    // Rename the subsection for the counters
+    OutStreamer->emitXCOFFRenameDirective(ProfCCsect->getQualNameSymbol(),
+                                          "__llvm_prf_cnts");
```
Can we use `getInstrProfSectionName` consistently everywhere rather than hardcoding the names as strings here? That gives us flexibility in case we need to change the name in the future.

mandlebug committed:
Add a new fixed metadata node to carry the xcoff pgo association.

Instead of decomposing the initializer for a __profd profiling data
symbol to get its associated __profc symbol attach a metadata node
tying the 2 respective globals together.

mandlebug committed:
Address review comments.

- Fixed spelling mistake in comment.
- Use getInstrProfSectionName instead of hardcoded section names.
- Fixup spacing in test to be consitently single spaced.

amy-kwan reviewed (approved):
Thanks for addressing the previous comments. I have no further comments on this PR, so LGTM.

mandlebug committed:
Convert cast<> to static cast.

MCSectionXCOFF no longer has the classof memeber function so it cannot
be used with cast<>.

mandlebug committed:
clang-format test source.

mandlebug committed:
Remove hardcoded use of "__llvm_prf_data."

mandlebug committed:
Revert "Remove hardcoded use of "__llvm_prf_data.""

This reverts commit a330ebd88eb40ad143b90f27978a9f8d03b91634.
The latest push failed on windows, and I'm guessing this may have
caused it.

amy-kwan reviewed (approved):
I think this LGTM.
